Remove TypeManager parameter from BlockEncodingFactory.readEncoding

haozhun opened the pull request:
The bulk of this PR is in "Remove hack to serialize a Block without using BlockEncodingSerde".

This makes option 2 in #10386 feasible.

findepi commented on presto-spi/src/test/java/com/facebook/presto/spi/block/TestingBlockEncodingSerde.java:
```diff
@@ -26,6 +26,8 @@
 import static java.nio.charset.StandardCharsets.UTF_8;
 import static java.util.Objects.requireNonNull;
 
+// This class is exactly the same as BlockEncodingManager.
```
How can we know this remains truth tomorrow?

a test of some kind?

haozhun replied:
I didn't add this. I just added a comment to explain the intention of the original author. I don't know how this can be guaranteed. I'm not sure how one can write such a test.

1. The other class is not visible in spi.
2. Even if it is, how does one assert that they are the same?

findepi replied:
If we mean that these classes are verbatim copies of each other, we could have a test that compares their sources (assuming tests are run from source tree). 
On higher level, we could check that they contain same set of encoding factories, etc.

findepi commented on presto-main/src/main/java/com/facebook/presto/block/BlockEncodingManager.java:
```diff
@@ -13,22 +13,22 @@
  */
 package com.facebook.presto.block;
 
-import com.facebook.presto.spi.block.ArrayBlockEncoding;
```
> Remove singleton instance of XBlockEncoding

Did you mean something like "`Remove singleton instances of block encodings`"?

findepi commented on presto-main/src/main/java/com/facebook/presto/block/BlockEncodingManager.java:
```diff
@@ -60,7 +59,7 @@ public BlockEncodingManager(TypeManager typeManager, Set<BlockEncodingFactory<?>
         // This function should be called from Guice and tests only
 
-        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        requireNonNull(typeManager, "typeManager is null");
```
Can be removed altogether, since `typeManager` is just passed through. Let `MapBlockEncodingFactory` and `SingleMapBlockEncodingFactory` validate that

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/planner/EffectivePredicateExtractor.java:
```diff
@@ -84,230 +83,234 @@ public static Expression extract(PlanNode node)
-        // Therefore, we can't say anything about the effective predicate of the
-        // output of such an aggregation.
-        if (node.getGroupingKeys().isEmpty()) {
+        @Override
```
Yes. You can make GitHub show it by appending `?w=1` to the end of the URI.

haozhun commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveFileFormats.java:
```diff
@@ -783,16 +788,16 @@ else if (testColumn.getObjectInspector().getCategory() == Category.PRIMITIVE) {
         }
     }
 
-    private static void assertBlockEquals(Block actual, Block expected, String message)
```
Let's pull all discussions about whether test field should be static in a separate place.

haozhun commented on presto-hive/src/test/java/com/facebook/presto/hive/util/TestSerDeUtils.java:
```diff
@@ -292,16 +297,16 @@ public void testReuse()
         assertBlockEquals(actual, expected);
     }
 
-    private static void assertBlockEquals(Block actual, Block expected)
```
Let's pull all discussions about whether test field should be static in a separate place.

haozhun commented on presto-main/src/test/java/com/facebook/presto/sql/TestSqlToRowExpressionTranslator.java:
```diff
@@ -46,36 +38,16 @@
-        metadata = null;
-    }
+    private final Metadata metadata = MetadataManager.createTestMetadataManager();
+    private final LiteralInterpreter literalInterpreter = new LiteralInterpreter(metadata.getBlockEncodingSerde());
```
Let's pull all discussions about whether test field should be static in a separate place.

haozhun commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestEffectivePredicateExtractor.java:
```diff
@@ -107,6 +109,9 @@
     private static final Expression GE = G.toSymbolReference();
 
+    private final Metadata metadata = MetadataManager.createTestMetadataManager();
+    private final EffectivePredicateExtractor effectivePredicateExtractor = new EffectivePredicateExtractor(new DomainTranslator(new LiteralInterpreter(metadata.getBlockEncodingSerde())));
```
Let's pull all discussions about whether test field should be static in a separate place.

haozhun commented on presto-main/src/test/java/com/facebook/presto/type/AbstractTestType.java:
```diff
@@ -52,6 +54,8 @@
 
 public abstract class AbstractTestType
 {
+    private final BlockEncodingSerde blockEncodingSerde = new BlockEncodingManager(new TypeRegistry());
```
Let's pull all discussions about whether test field should be static in a separate place.

haozhun commented:
Regarding whether constant fields in tests should be static. The existing code is pretty inconsistent. You can find that by searching `createTestingMetadataManager`. Most are not static, but there exists ones that are.

I chatted about this topic at breakfast. There isn't a strong preference here. @electrum pointed out that it looks better that those fields are not capitalized.

There isn't much of a real difference to lead to a strong preference:
* Testing classes are not generally initialized more than once.
* Testing methods are not static. (Otherwise, static field will be preferred.)
* Fields are generally not expensive enough that garbage collecting them matters. (Maybe they can't be garbage collected anyways because TestNG hold on to them.)

I notice that existing code is not consistent on this choice. As a result, when I worked on this PR, I made the choice based on which choice the class made previously. When choice has not been established in a class, I chose non static.

I'll keep this as is unless you have a strong preference.

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralInterpreter.java:
```diff
@@ -109,7 +115,16 @@ public static Object evaluate(Metadata metadata, ConnectorSession session, Expre
     }
 
-    public static Expression toExpression(Object object, Type type)
+    public static Expression bigintLiteral(Object object)
```
inlined back

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java:
```diff
@@ -225,9 +220,10 @@ public static void verifyExpressionIsConstant(Set<NodeRef<Expression>> columnRef
-        this.session = session.toConnectorSession();
+        this.expression = requireNonNull(expression, "expression is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.literalInterpreter = new LiteralInterpreter(metadata.getBlockEncodingSerde());
```
I don't have a strong opinion on this. For this, a separate commit will probably be easier to work with.

findepi replied:
@haozhun yes, if you go this route, a separate commit is a good idea.

@martint we're considering splitting `LiteralInterpreter` into two classes here (or renaming it?). Do you have an opinion?

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java:
```diff
@@ -655,15 +651,15 @@ else if (!found && (Boolean) invokeOperator(OperatorType.EQUAL, types(node.getVa
                                 .filter((expression -> !DeterminismEvaluator.isDeterministic(expression))))
                         .collect(toImmutableList());
-                return new InPredicate(toExpression(value, type), new InListExpression(simplifiedExpressionValues));
+                return new InPredicate(literalInterpreter.toExpression(value, type), new InListExpression(simplifiedExpressionValues));
```
Done. (I can see arguments either way. I don't have a strong opinion here.)

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java:
```diff
@@ -176,11 +183,11 @@ public Result apply(TableScanNode tableScanNode, Captures captures, Context cont
     }
 
-    private static PlanNode planTableScan(TableScanNode node, Expression predicate, Rule.Context context, Metadata metadata)
+    private static PlanNode planTableScan(TableScanNode node, Expression predicate, Rule.Context context, Metadata metadata, DomainTranslator domainTranslator)
```
Done

haozhun commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java:
```diff
@@ -1052,12 +1053,12 @@ private void testNumericTypeTranslation(NumericValues columnValues, NumericValue
+        Expression min = LITERAL_INTERPRETER.toExpression(literalValues.getMin(), literalType);
+        Expression integerPositive = LITERAL_INTERPRETER.toExpression(literalValues.getIntegerPositive(), literalType);
+        Expression integerNegative = LITERAL_INTERPRETER.toExpression(literalValues.getIntegerNegative(), literalType);
+        Expression fractionalPositive = LITERAL_INTERPRETER.toExpression(literalValues.getFractionalPositive(), literalType);
```
Done

findepi commented:
> Regarding whether constant fields in tests should be static.  [....]

I noticed the existing inconsistence. I agree that `STATIC_FIELDS` are less readable than `instanceFields` and -- as you noticed, due to how testng works -- there is no difference in retention. I believe that static fields serve slightly as a reminder that tests are concurrent. E.g. we can have `@BeforeClass` initializing some static state, but we should have no `@BeforeMethod` methods initializing instance fields (unless a class is marked single-threaded).

> I'll keep this as is unless you have a strong preference.

👍  , i don't see a need to fix/change all this now

haozhun commented:
I addressed comments from @findepi. 

There's still an outstanding discussion about splitting LiteralInterpreter, which can potentially be a separate commit or separate PR if we choose to do it.

haozhun commented:
Thank you for your suggestion, @findepi. I talked with @martint and we decided to introduce a `LiteralEncoder`.

@findepi Can you review this PR again and approve (if appropriate)?

haozhun commented:
I discussed with @electrum in detail for the interface of replacement block part of this PR. I'm going ahead with merging this PR.
